txp/Pos/Txp/Toil/Types.hs

```diff
@@ -124,18 +126,39 @@ instance Default MemPool where
         }
 
 ----------------------------------------------------------------------------
--- ToilModifier
+-- UtxoModifier, UndoMap and AddrCoinsMap
 ----------------------------------------------------------------------------
 
 type UtxoModifier = MM.MapModifier TxIn TxOutAux
 type UndoMap = HashMap TxId TxUndo
-
-fromUtxo :: Utxo -> UtxoModifier
-fromUtxo = foldr (uncurry MM.insert) mempty . M.toList
+type AddrCoinMap = HashMap Address Coin
+
+utxoToModifier :: Utxo -> UtxoModifier
+utxoToModifier = foldr (uncurry MM.insert) mempty . M.toList
+
+-- | Takes utxo modifier and address-coin map with correspodning utxo
+-- and applies utxo modifier to map.
+-- Works for O(size of modifier * log (size of map)).
+applyUtxoModToAddrCoinMap
+    :: UtxoModifier
+    -> (AddrCoinMap, Utxo)
+    -> AddrCoinMap
+applyUtxoModToAddrCoinMap modifier (addrCoins, utxo) =
+    let outToAddr = txOutAddress . toaOut in
+    let resolvedAddrs =
+          mapMaybe (fmap outToAddr . flip M.lookup utxo)
+                   (MM.deletions modifier) in
+    let addrCoinsRest = foldr HM.delete addrCoins resolvedAddrs in
+    let modAddrCoins = map (view _TxOut . toaOut . snd) (MM.insertions modifier) in
+    foldr (uncurry $ HM.insertWith unsafeAddCoin) addrCoinsRest modAddrCoins
```

Comment:
This is incorrect. What do you do here:
1) Take list of all addresses which have at least one output spent, according to `UtxoModifier`
2) _Delete_ all this addresses from `AddrCoinMap`
3) For every newly added unspent output in `UtxoModifier`, add its value to corresponding address in `AddrCoinMap`
In (2) there is a bug - there might be more than one output with money on the address, and not all outputs might be spent in given `UtxoModifier`. Yes, in regular circumstances this won't be the case - because our own tx creation logic spend _all_ money on address every time, but we cannot guarantee that transactions which spend money on address only partially won't appear on blockchain. So this needs to be fixed.